core/arch/arm/kernel/mutex.c

```diff
@@ -91,37 +91,118 @@ static void __mutex_unlock(struct mutex *m, const char *fname, int lineno)
 
 	old_itr_status = cpu_spin_lock_xsave(&m->spin_lock);
 
-	if (m->value != MUTEX_VALUE_LOCKED)
+	if (!m->state)
 		panic();
 
 	thread_rem_mutex(m);
-	m->value = MUTEX_VALUE_UNLOCKED;
+	m->state = 0;
 
 	cpu_spin_unlock_xrestore(&m->spin_lock, old_itr_status);
 
-	wq_wake_one(&m->wq, m, fname, lineno);
+	wq_wake_next(&m->wq, m, fname, lineno);
 }
 
 static bool __mutex_trylock(struct mutex *m, const char *fname __unused,
 			int lineno __unused)
 {
 	uint32_t old_itr_status;
-	enum mutex_value old_value;
+	bool can_lock;
```

Comment:
`can_lock_write` or similar will be more descriptive. I'm looking from new developer PoV.